extensions/Rules/lib/Rule.pm

```diff
@@ -0,0 +1,212 @@
+
+package Bugzilla::Extension::Rules::Rule;
+
+use 5.10.1;
+use Data::Dumper;
+use Moo;
+use List::Util qw(any none);
+use Types::Standard -all;
+use Type::Utils;
+
+use Bugzilla::Logging;
+use Bugzilla::Status;
+use Bugzilla::Types qw(Bug FakeBug User);
+
+#########################
+#    Initialization     #
+#########################
+
+# Rule definitions
+has action    => (is => 'ro', isa => ArrayRef | Str);
+has change    => (is => 'ro', isa => Maybe [Dict]);
+has condition => (is => 'ro', isa => Maybe [Dict]);
+has error     => (is => 'ro', isa => Maybe [Str]);
+has filter    => (is => 'ro', isa => Maybe [Dict]);
+has name      => (is => 'ro', isa => Str);
+
+# Current attributes
+has bug       => (is => 'ro', isa => Bug | FakeBug);
+has user      => (is => 'ro', isa => User);
+has field     => (is => 'ro', isa => Str);
+has old_value => (is => 'ro', isa => Str);
+has new_value => (is => 'ro', isa => Str);
+
+sub BUILDARGS {
+  my ($class, $params) = @_;
+
+  $params->{action}    = $params->{rule}->{action};
+  $params->{change}    = $params->{rule}->{change};
+  $params->{condition} = $params->{rule}->{condition};
+  $params->{error}     = $params->{rule}->{error};
+  $params->{filter}    = $params->{rule}->{filter};
+  $params->{name}      = $params->{rule}->{name};
+
+  delete $params->{rule};
+
+  return $params;
+}
+
+#########################
+#        Actions        #
+#########################
+
+sub process {
+  my ($self) = @_;
+  my $matches = [];
+
+  # $self->_debug_info();
+
+  # Process filters to match this rule to the bugs current attributes
+  $self->_process_filters($matches);
+
+  # Process what is changing and what the new/old values are
+  $self->_process_changes($matches);
+
+  # Process any special conditions that need to be met
+  $self->_process_conditions($matches);
+
+  # If we have not fully matched by this point we ignore this rule
+  if (any { $_ == 0 } @{$matches}) {
+    # DEBUG('NO MATCH');
+    return {name => $self->name, action => 'none'};
+  }
+  else {
+    my $result = {name => $self->name, action => 'allow'};
+
+    # Process actions since we matched
+    my $actions = ref $self->action ? $self->action : [$self->action];
+
+    # cannot_create means we disallow this change for new bugs
+    if (any { $_ eq 'cannot_create' } @{$actions} && !$self->bug->id) {
+      $result->{action} = 'deny';
+    }
+
+    # cannot_update means we disallow this change for any bug, even current
+    if (any { $_ eq 'cannot_update' || $_ eq 'cannot_comment' } @{$actions}) {
+      $result->{action} = 'deny';
+    }
+
+    $result->{error} = $self->error if $result->{action} eq 'deny';
+
+    # DEBUG('MATCHED: ' . $result->{action});
+
+    return $result;
+  }
+}
+
+#############################
+#      Privcate Methods     #
+#############################
+
+sub _process_filters {
+  my ($self, $matches) = @_;
+
+  if (my $filter = $self->filter) {
+    foreach my $item (qw(product component)) {
```

Comment:
for most (all?) of the filters, changes, and conditions it would be must more convenient if they were case insensitive matches.